chainer/function_node.py

```diff
@@ -0,0 +1,591 @@
+         backpropagations so that the automatic higher-order differentiation is
+         available.
+       - The backpropagation of the new-style differentiable function can be
+         more computationally saving because the interface allows an
+         implementation to omit the computation of unneeded input gradients.
+
+       Note that the new-style differentiable function is the standard way of
+       defining a function node of the computational graph in Chainer; old-
+       style differentiable functions are implemented as wrappers of the new-
+       style differentiable functions.
+
+    Attributes:
+        inputs: A tuple of the input :class:`VariableNode` objects.
+        outputs: A tuple of weak references to the output
+            :class:`VariableNode` objects.
+        rank (int): An ordinal following the topological order of the
+            computational graph.
+        stack: Stack trace retrieved at the forward computation. The stack
+            trace is available only in the debug mode.
+
+    .. versionadded:: 3.0.0
+
+    """
+
+    inputs = None
+    outputs = None
+    rank = 0
+    stack = None
+    _input_indexes_to_retain = None
+    _output_indexes_to_retain = None
+    _retained_output_data = None
+    _local_function_hooks = None
+
+    @property
+    def local_function_hooks(self):
+        """Ordered dictionary of registered function hooks.
+
+        Contrary to ``chainer.thread_local.function_hooks``,
+        which registers its elements to all functions,
+        Function hooks in this property is specific to this function.
+
+        """
+        if self._local_function_hooks is None:
+            self._local_function_hooks = collections.OrderedDict()
+        return self._local_function_hooks
+
+    @property
+    def _n_local_function_hooks(self):
+        return (0 if self._local_function_hooks is None
+                else len(self._local_function_hooks))
+
+    @property
+    def label(self):
+        """Short text that represents the function.
+
+        The default implementation returns its type name.
+        Each function should override it to give more information.
+
+        """
+        return self.__class__.__name__
+
+    @property
+    def output_data(self):
+        """A tuple of the retained output arrays.
+
+        This property is mainly used by :class:`Function`. Users basically do
+        not have to use this property; use :meth:`get_retained_outputs`
+        instead.
+
+        """
+        if self._retained_output_data is None:
+            raise RuntimeError('retained output data is gone')
+        out_data = [None] * len(self.outputs)
+        for index, data in six.moves.zip(self._output_indexes_to_retain,
+                                         self._retained_output_data):
+            out_data[index] = data
+        return tuple(out_data)
+
+    @property
+    def _impl_name(self):
+        return self.__class__.__name__
+
+    def apply(self, inputs):
```

Comment:
I thought that passing a sequence directly is more verbose and sometimes makes the code simpler (esp. in the case the caller also builds the arguments as a tuple or a list). It is also slightly faster than `*inputs` (25% faster when `inputs` is a tuple, and 100% faster when `inputs` is a list in my env). Is it confusing to use a calling convention different from `Function`?